domain/keymanager/modelmigration/import.go

```diff
@@ -0,0 +1,232 @@
+		userName, err := user.NewName(uak.Username())
+		if err != nil {
+			return errors.Errorf(
+				"cannot import authorized keys for user %q on model when constructing user name: %w",
+				uak.Username(), err,
+			)
+		}
+
+		user, err := i.userService.GetUserByName(ctx, userName)
+		if errors.Is(err, accesserrors.UserNotFound) {
+			return errors.Errorf(
+				"cannot import authorized keys for user %q on model, user does not exist in the model",
+				userName,
+			).Add(err)
+		} else if err != nil {
+			return errors.Errorf(
+				"cannot import authorized keys for user %q on model when finding user: %w",
+				userName,
+				err,
+			)
+		}
+
+		modelUUID := coremodel.UUID(model.Tag().Id())
+		if err := modelUUID.Validate(); err != nil {
+			return errors.Errorf(
+				"importing authorized keys for model %q: %w", modelUUID, err,
+			)
+		}
+
+		err = i.serviceGetter(modelUUID).AddPublicKeysForUser(ctx, user.UUID, uak.AuthorizedKeys()...)
+		if err != nil {
+			return errors.Errorf(
+				"cannot import authorized keys for user %q on model: %w",
+				userName,
+				err,
+			)
+		}
+	}
+
+	return nil
+}
+
+// executeModelConfigAuthorizedKeys is responsible for importing a models
+// authorized keys when they are still contained with the models config. When we
+// detect that we are importing a model that has still been storing authorized
+// keys within model config we want to pull these keys out and import them into
+// the model under the admin user.
+//
+// NOTE (tlm): It was chosen to perform this logic and action here instead of in
+// the description package between description versions because it would have
+// required the description package to start getting internal business logic
+// about what user owns what key.
+func (i *importOperation) executeModelConfigAuthorizedKeys(
+	ctx context.Context,
+	model description.Model,
+) error {
+	authKeysI, has := model.Config()[modelConfigKeyAuthorizeKeys]
+	if !has {
+		// No authorized keys in model config so we can safely just get out of
+		// here.
+		return nil
+	}
+	authKeys, isString := authKeysI.(string)
+	if !isString {
+		return errors.New("cannot import authorized keys from model config, expected a string")
+	}
+
+	// Because of bugs over time there are Juju controllers out in the wild that
+	// have the controllers ssh key baked into each and every models authorized
+	// keys. We can't fix this now as the damage is done.
+	//
+	// But we can do our best to stop it coming across when importing models. It
+	// would be a security bug if an old controller could still access the
+	// machines of a model that it does not own anymore.
+	publicKeys, err := ssh.SplitAuthorizedKeys(authKeys)
+	if err != nil {
+		return errors.Errorf(
+			"cannot split authorized keys in model config during import: %w",
+			err,
+		)
+	}
+
+	cleansedKeys := make([]string, 0, len(publicKeys))
+	for _, pk := range publicKeys {
+		if strings.Contains(pk, controller.ControllerSSHKeyComment) {
+			continue
```

Comment:
Explicitly warning log here that we are throwing away data.

Reply:
Done & Done but I think warning is a bit much. I would personally go debug here.